File: php/src/DomainClients.php
<?php

declare(strict_types=1);

namespace TwoFinance\Sdk;

final class AuthClient
{
    public function __construct(
        private readonly ServiceClient $service,
        private readonly string $realm = '2finance',
        private readonly string $clientId = '2finance-network',
        private readonly string $phoneClientId = '2finance-network-phone',
    ) {
    }

    /**
     * @return mixed
     */
    public function login(array $input): mixed
    {
        return $this->service->post($this->authPath($this->clientId, '/login'), $input);
    }

    /**
     * @return mixed
     */
    public function refreshToken(string $refreshToken): mixed
    {
        return $this->service->post($this->authPath($this->clientId, '/refresh'), ['refresh_token' => $refreshToken]);
    }

    /**
     * @return mixed
     */
    public function phoneLogin(string $phoneNumber, string $code): mixed
    {
        return $this->service->post(
            $this->authPath($this->phoneClientId, '/phone/sms/login'),
            ['phone_number' => $phoneNumber, 'code' => $code],
        );
    }

    public function jwks(): mixed
    {
        return $this->service->get($this->oidcPath('/protocol/openid-connect/certs'));
    }

    public function validateToken(string $token): mixed
    {
        return $this->service->post($this->oidcPath('/protocol/openid-connect/token/introspect'), ['token' => $token]);
    }

    private function authPath(string $clientId, string $endpoint): string
    {
        return sprintf('/v1/2finance-authenticator/%s/%s/%s', $this->realm, $clientId, ltrim($endpoint, '/'));
    }

    private function oidcPath(string $endpoint): string
    {
        return sprintf('/realms/%s/%s', $this->realm, ltrim($endpoint, '/'));
    }
}

final class AnalyticsClient
{
    public function __construct(private readonly ServiceClient $service)
    {
    }

    public function indicators(): mixed
    {
        return $this->service->get('/analytics/indicators');
    }

    public function get(string $path, ?RequestOptions $options = null): mixed
    {
        return $this->service->get($path, $options);
    }

    public function post(string $path, mixed $body = null, ?RequestOptions $options = null): mixed
    {
        return $this->service->post($path, $body, $options);
    }

    public function calculateTechnicalAnalysis(array $request): mixed
    {
        return $this->service->post('/analytics/technical-analysis:calculate', $request);
    }

    public function optimizePortfolio(array $request): mixed
    {
        return $this->service->post('/portfolio-manager/optimizer', $request);
    }

    public function upsertCandles(array $request): mixed
    {
        return $this->service->post('/analytics/candles:upsert', $request);
    }

    public function rankings(): mixed
    {
        return $this->service->get('/portfolio-manager/rankings');
    }

    public function balances(string $accountId): mixed
    {
        return $this->service->get('/portfolio-manager/balances/' . rawurlencode($accountId));
    }

    public function blackScholes(string $query = ''): mixed
    {
        return $this->service->get('/risk-manager/blackscholes' . ($query === '' ? '' : '?' . $query));
    }

    public function staking(): mixed
    {
        return $this->service->get('/staking');
    }
}

final class MCPClient
{
    private int $nextId = 1;

    public function __construct(private readonly ServiceClient $service)
    {
    }

    public function call(string $method, mixed $params = null): mixed
    {
        return $this->service->post('/mcp', [
            'jsonrpc' => '2.0',
            'id' => $this->nextId++,
            'method' => $method,
            'params' => $params,
        ]);
    }

    public function listTools(): mixed
    {
        return $this->call('tools/list');
    }

    public function listPrompts(): mixed
    {
        return $this->call('prompts/list');
    }

    public function listResources(): mixed
    {
        return $this->call('resources/list');
    }

    public function readResource(string $uri): mixed
    {
        return $this->call('resources/read', ['uri' => $uri]);
    }

    public function callTool(string $name, array $arguments = []): mixed
    {
        return $this->call('tools/call', ['name' => $name, 'arguments' => $arguments]);
    }

    public function getPrompt(string $name, array $arguments = []): mixed
    {
        return $this->call('prompts/get', ['name' => $name, 'arguments' => $arguments]);
    }

    public function conversationPlan(array $arguments): mixed
    {
        return $this->callTool('finance_assistant.conversation.plan', $arguments);
    }
}

final class OrchestratorClient
{
    public function __construct(private readonly ServiceClient $service)
    {
    }

    public function catalog(): mixed
    {
        return $this->service->get('/v1/mcphost/catalog/packages');
    }

    public function createSession(array $request): mixed
    {
        return $this->service->post('/v1/mcphost/sessions', $request);
    }

    public function sendMessage(array $request): mixed
    {
        return $this->service->post('/v1/mcphost/messages', $request);
    }

    public function tools(): mixed
    {
        return $this->service->get('/v1/mcphost/tools');
    }

    public function prompts(): mixed
    {
        return $this->service->get('/v1/mcphost/prompts');
    }

    public function resources(): mixed
    {
        return $this->service->get('/v1/mcphost/resources');
    }

    public function providers(): mixed
    {
        return $this->service->get('/v1/mcphost/providers');
    }

    public function approvals(): mixed
    {
        return $this->service->get('/v1/mcphost/approvals');
    }

    public function observability(): mixed
    {
        return $this->service->get('/v1/mcphost/observability');
    }

    public function deleteSession(string $id): mixed
    {
        return $this->service->delete('/v1/mcphost/sessions/' . rawurlencode($id));
    }
}

final class NetworkClient
{
    public function __construct(private readonly ServiceClient $service)
    {
    }

    public function virtualMachine(): mixed
    {
        return $this->service->post('/v2/2finance-network/query', [
            'method' => 'get_blocks',
            'params' => ['page' => 1, 'limit' => 1],
        ]);
    }

    public function marketCandles(string $market, string $query = ''): mixed
    {
        return $this->service->get(
            '/v2/2finance-network/markets/' . rawurlencode($market) . '/candles' . ($query === '' ? '' : '?' . $query),
        );
    }

    public function products(string $productType): mixed
    {
        return $this->service->get('/v2/2finance-network/products/' . rawurlencode($productType));
    }

    public function createProduct(string $productType, array $request): mixed
    {
        return $this->service->post('/v2/2finance-network/products/' . rawurlencode($productType), $request);
    }

    public function bonds(): mixed
    {
        return $this->products('bonds');
    }

    public function createBond(array $request): mixed
    {
        return $this->createProduct('bonds', $request);
    }

    public function loans(): mixed
    {
        return $this->products('loans');
    }

    public function createLoan(array $request): mixed
    {
        return $this->createProduct('loans', $request);
    }

    public function swaps(): mixed
    {
        return $this->products('swaps');
    }

    public function createSwap(array $request): mixed
    {
        return $this->createProduct('swaps', $request);
    }

    public function stakingProducts(): mixed
    {
        return $this->products('staking');
    }

    public function createStakingProduct(array $request): mixed
    {
        return $this->createProduct('staking', $request);
    }

    public function syntheticAssets(): mixed
    {
        return $this->products('synthetic-assets');
    }

    public function createSyntheticAsset(array $request): mixed
    {
        return $this->createProduct('synthetic-assets', $request);
    }

    public function liquidityPools(): mixed
    {
        return $this->products('liquidity-pools');
    }

    public function createLiquidityPool(array $request): mixed
    {
        return $this->createProduct('liquidity-pools', $request);
    }
}

final class TradingControlClient
{
    public function __construct(private readonly ServiceClient $service)
    {
    }

    public function robots(): mixed
    {
        return $this->service->get('/robots');
    }

    public function createRobot(array $request): mixed
    {
        return $this->service->post('/robots', $request);
    }

    public function robot(string $id): mixed
    {
        return $this->service->get('/robots/' . rawurlencode($id));
    }

    public function startRobot(string $id): mixed
    {
        return $this->service->post('/robots/' . rawurlencode($id) . ':start');
    }

    public function pauseRobot(string $id): mixed
    {
        return $this->service->post('/robots/' . rawurlencode($id) . ':pause');
    }

    public function resumeRobot(string $id): mixed
    {
        return $this->service->post('/robots/' . rawurlencode($id) . ':resume');
    }

    public function stopRobot(string $id): mixed
    {
        return $this->service->post('/robots/' . rawurlencode($id) . ':stop');
    }

    public function riskPolicy(string $id): mixed
    {
        return $this->service->get('/robots/' . rawurlencode($id) . '/risk-policy');
    }

    public function setRiskPolicy(string $id, array $request): mixed
    {
        return $this->service->put('/robots/' . rawurlencode($id) . '/risk-policy', $request);
    }

    public function riskView(string $id): mixed
    {
        return $this->service->get('/risk-view/' . rawurlencode($id));
    }

    public function strategies(): mixed
    {
        return $this->service->get('/strategies');
    }

    public function createStrategy(array $request): mixed
    {
        return $this->service->post('/strategies', $request);
    }

    public function directives(): mixed
    {
        return $this->service->get('/directives');
    }

    public function createDirective(array $request): mixed
    {
        return $this->service->post('/directives', $request);
    }

    public function audit(): mixed
    {
        return $this->service->get('/audit');
    }

    public function activity(): mixed
    {
        return $this->service->get('/activity');
    }

    public function mcpTools(): mixed
    {
        return $this->service->get('/mcp/tools');
    }
}

final class KeyStoreClient
{
    public function __construct(private readonly ServiceClient $service)
    {
    }

    public function health(): mixed
    {
        return $this->service->get('/healthz');
    }

    public function readiness(): mixed
    {
        return $this->service->get('/readyz');
    }

    public function startKeygen(array $request): mixed
    {
        return $this->service->post('/keystore/keygen/start', $request);
    }

    public function keygenSignature(array $request): mixed
    {
        return $this->service->post('/keystore/keygen/signature', $request);
    }

    public function startSigning(array $request): mixed
    {
        return $this->service->post('/keystore/signing/start', $request);
    }

    public function signingSignature(array $request): mixed
    {
        return $this->service->post('/keystore/signing/signature', $request);
    }

    public function startResharing(array $request): mixed
    {
        return $this->service->post('/keystore/resharing/start', $request);
    }

    public function keys(string $userPublicKey): mixed
    {
        return $this->service->get('/keystore/keys/' . rawurlencode($userPublicKey));
    }

    public function signatures(string $userPublicKey): mixed
    {
        return $this->service->get('/keystore/signatures/' . rawurlencode($userPublicKey));
    }

    public function metrics(): mixed
    {
        return $this->service->get('/keystore/tss/metrics');
    }
}

final class HummingbotClient
{
    public function __construct(private readonly ServiceClient $service)
    {
    }

    public function assets(): mixed
    {
        return $this->service->get('/api/v1/assets');
    }

    public function symbols(): mixed
    {
        return $this->service->get('/api/v1/symbols');
    }

    public function balances(): mixed
    {
        return $this->service->get('/api/v1/balances');
    }

    public function connectorConfig(array $request): mixed
    {
        return $this->service->post('/api/v1/connectors/2finance/config', $request);
    }
}

final class ProviderClient
{
    public function __construct(private readonly ServiceClient $service)
    {
    }

    public function get(string $path): mixed
    {
        return $this->service->get($path);
    }

    public function post(string $path, mixed $body = null): mixed
    {
        return $this->service->post($path, $body);
    }

    public function put(string $path, mixed $body = null): mixed
    {
        return $this->service->put($path, $body);
    }

    public function delete(string $path): mixed
    {
        return $this->service->delete($path);
    }
}

final class WiseClient
{
    public function __construct(private readonly ProviderClient $provider)
    {
    }

    public function get(string $path): mixed
    {
        return $this->provider->get($path);
    }

    public function post(string $path, mixed $body = null): mixed
    {
        return $this->provider->post($path, $body);
    }

    public function profiles(): mixed
    {
        return $this->get('/v1/profiles');
    }

    public function profile(string $profileId): mixed
    {
        return $this->get('/v1/profiles/' . rawurlencode($profileId));
    }

    public function createQuote(string $profileId, array $request): mixed
    {
        return $this->post('/v3/profiles/' . rawurlencode($profileId) . '/quotes', $request);
    }

    public function createTransfer(array $request): mixed
    {
        return $this->post('/v1/transfers', $request);
    }
}

final class AirwallexClient
{
    public function __construct(private readonly ProviderClient $provider)
    {
    }

    public function get(string $path): mixed
    {
        return $this->provider->get($path);
    }

    public function post(string $path, mixed $body = null): mixed
    {
        return $this->provider->post($path, $body);
    }

    public function accounts(): mixed
    {
        return $this->get('/api/v1/accounts');
    }

    public function payments(): mixed
    {
        return $this->get('/api/v1/payments');
    }

    public function createPayment(array $request): mixed
    {
        return $this->post('/api/v1/payments', $request);
    }

    public function beneficiaries(): mixed
    {
        return $this->get('/api/v1/beneficiaries');
    }

    public function createBeneficiary(array $request): mixed
    {
        return $this->post('/api/v1/beneficiaries', $request);
    }
}

final class MatchEngineClient
{
    public function __construct(public readonly string $webSocketUrl)
    {
    }

    /**
     * @param array<string,mixed> $command
     * @return array<string,mixed>
     */
    public function orderCommand(array $command): array
    {
        return array_merge([
            'schema' => 'matchengine.order_command.v2',
            'message_type' => 'ORDER',
            'operation' => 'ADD',
        ], $command);
    }

    /**
     * @param array<string,mixed> $request
     * @return array<string,mixed>
     */
    public function marketDataSubscribe(array $request): array
    {
        return ['schema' => 'matchengine.market_data_subscribe.v2'] + $request;
    }

    public function sendOrder(callable $sender, array $command): mixed
    {
        return $sender($this->orderCommand($command));
    }

    public function subscribeMarketData(callable $sender, array $request): mixed
    {
        return $sender($this->marketDataSubscribe($request));
    }
}

final class PlannerClient
{
    public function __construct(
        private readonly MCPClient $mcp,
        private readonly OrchestratorClient $orchestrator,
        private readonly AnalyticsClient $analytics,
        private readonly TradingControlClient $tradingControl,
    ) {
    }

    public function conversationPlan(array $request): mixed
    {
        return $this->mcp->conversationPlan($request);
    }

    public function orchestratedPlan(array $request): mixed
    {
        return $this->orchestrator->sendMessage($request);
    }

    public function operationalPlan(array $request): mixed
    {
        return $this->orchestratedPlan($request);
    }

    public function tradingPlan(array $request): mixed
    {
        $context = is_array($request['context'] ?? null) ? $request['context'] : [];
        if (($request['use_trading'] ?? false) === true) {
            try {
                $context['trading_robots'] = $this->tradingControl->robots();
            } catch (\Throwable) {
                // Best-effort enrichment keeps planning usable when trading is unavailable.
            }
        }
        if (($request['use_analytics'] ?? false) === true) {
            try {
                $context['analytics_indicators'] = $this->analytics->indicators();
            } catch (\Throwable) {
                // Best-effort enrichment keeps planning usable when analytics is unavailable.
            }
        }
        $request['context'] = $context;
        return $this->conversationPlan($request);
    }
}

File: java/src/main/java/com/twofinance/sdk/DomainClients.java
package com.twofinance.sdk;

import java.io.IOException;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.nio.charset.StandardCharsets;
import java.util.function.Function;

public final class DomainClients {
    private DomainClients() {}

    public static class AuthClient extends ServiceClient {
        private final String realm;
        private final String clientId;
        private final String phoneClientId;

        public AuthClient(String baseUrl, HttpClient httpClient, TokenSource tokenSource, SDKConfig config) {
            super(baseUrl, httpClient, tokenSource);
            this.realm = config.authRealm;
            this.clientId = config.authClientId;
            this.phoneClientId = config.authPhoneClientId;
        }

        public String login(String jsonBody) throws IOException, InterruptedException {
            return post(authPath(clientId, "/login"), jsonBody);
        }

        public String refreshToken(String jsonBody) throws IOException, InterruptedException {
            return post(authPath(clientId, "/refresh"), jsonBody);
        }

        public String phoneLogin(String jsonBody) throws IOException, InterruptedException {
            return post(authPath(phoneClientId, "/phone/sms/login"), jsonBody);
        }

        public String jwks() throws IOException, InterruptedException {
            return get(oidcPath("/protocol/openid-connect/certs"));
        }

        public String validateToken(String jsonBody) throws IOException, InterruptedException {
            return post(oidcPath("/protocol/openid-connect/token/introspect"), jsonBody);
        }

        private String authPath(String selectedClientId, String endpoint) {
            return "/v1/2finance-authenticator/" + realm + "/" + selectedClientId + "/" + endpoint.replaceFirst("^/+", "");
        }

        private String oidcPath(String endpoint) {
            return "/realms/" + realm + "/" + endpoint.replaceFirst("^/+", "");
        }
    }

    public static class AnalyticsClient extends ServiceClient {
        public AnalyticsClient(String baseUrl, HttpClient httpClient, TokenSource tokenSource) {
            super(baseUrl, httpClient, tokenSource);
        }

        public String indicators() throws IOException, InterruptedException {
            return get("/analytics/indicators");
        }

        public String calculateTechnicalAnalysis(String jsonBody) throws IOException, InterruptedException {
            return post("/analytics/technical-analysis:calculate", jsonBody);
        }

        public String optimizePortfolio(String jsonBody) throws IOException, InterruptedException {
            return post("/portfolio-manager/optimizer", jsonBody);
        }

        public String upsertCandles(String jsonBody) throws IOException, InterruptedException {
            return post("/analytics/candles:upsert", jsonBody);
        }

        public String rankings() throws IOException, InterruptedException {
            return get("/portfolio-manager/rankings");
        }

        public String balances(String accountId) throws IOException, InterruptedException {
            return get("/portfolio-manager/balances/" + encode(accountId));
        }

        public String blackScholes(String query) throws IOException, InterruptedException {
            return get("/risk-manager/blackscholes" + (query == null || query.isBlank() ? "" : "?" + query));
        }

        public String staking() throws IOException, InterruptedException {
            return get("/staking");
        }
    }

    public static class MCPClient extends ServiceClient {
        private long nextId = 1;

        public MCPClient(String baseUrl, HttpClient httpClient, TokenSource tokenSource) {
            super(baseUrl, httpClient, tokenSource);
        }

        public String call(String method, String paramsJson) throws IOException, InterruptedException {
            String params = paramsJson == null || paramsJson.isBlank() ? "null" : paramsJson;
            String body = "{\"jsonrpc\":\"2.0\",\"id\":" + nextId++ + ",\"method\":\"" + method + "\",\"params\":" + params + "}";
            return post("/mcp", body);
        }

        public String conversationPlan(String argumentsJson) throws IOException, InterruptedException {
            String args = argumentsJson == null || argumentsJson.isBlank() ? "{}" : argumentsJson;
            return call("tools/call", "{\"name\":\"finance_assistant.conversation.plan\",\"arguments\":" + args + "}");
        }

        public String listTools() throws IOException, InterruptedException {
            return call("tools/list", null);
        }

        public String listPrompts() throws IOException, InterruptedException {
            return call("prompts/list", null);
        }

        public String listResources() throws IOException, InterruptedException {
            return call("resources/list", null);
        }

        public String readResource(String uri) throws IOException, InterruptedException {
            return call("resources/read", "{\"uri\":\"" + escape(uri) + "\"}");
        }

        public String callTool(String name, String argumentsJson) throws IOException, InterruptedException {
            String args = argumentsJson == null || argumentsJson.isBlank() ? "{}" : argumentsJson;
            return call("tools/call", "{\"name\":\"" + escape(name) + "\",\"arguments\":" + args + "}");
        }

        public String getPrompt(String name, String argumentsJson) throws IOException, InterruptedException {
            String args = argumentsJson == null || argumentsJson.isBlank() ? "{}" : argumentsJson;
            return call("prompts/get", "{\"name\":\"" + escape(name) + "\",\"arguments\":" + args + "}");
        }
    }

    public static class OrchestratorClient extends ServiceClient {
        public OrchestratorClient(String baseUrl, HttpClient httpClient, TokenSource tokenSource) {
            super(baseUrl, httpClient, tokenSource);
        }

        public String catalog() throws IOException, InterruptedException {
            return get("/v1/mcphost/catalog/packages");
        }

        public String createSession(String jsonBody) throws IOException, InterruptedException {
            return post("/v1/mcphost/sessions", jsonBody);
        }

        public String sendMessage(String jsonBody) throws IOException, InterruptedException {
            return post("/v1/mcphost/messages", jsonBody);
        }

        public String tools() throws IOException, InterruptedException {
            return get("/v1/mcphost/tools");
        }

        public String prompts() throws IOException, InterruptedException {
            return get("/v1/mcphost/prompts");
        }

        public String resources() throws IOException, InterruptedException {
            return get("/v1/mcphost/resources");
        }

        public String providers() throws IOException, InterruptedException {
            return get("/v1/mcphost/providers");
        }

        public String approvals() throws IOException, InterruptedException {
            return get("/v1/mcphost/approvals");
        }

        public String observability() throws IOException, InterruptedException {
            return get("/v1/mcphost/observability");
        }

        public String deleteSession(String id) throws IOException, InterruptedException {
            return delete("/v1/mcphost/sessions/" + encode(id));
        }
    }

    public static class NetworkClient extends ServiceClient {
        public NetworkClient(String baseUrl, HttpClient httpClient, TokenSource tokenSource) {
            super(baseUrl, httpClient, tokenSource);
        }

        public String virtualMachine() throws IOException, InterruptedException {
            return post(
                "/v2/2finance-network/query",
                "{\"method\":\"get_blocks\",\"params\":{\"page\":1,\"limit\":1}}"
            );
        }

        public String marketCandles(String market, String query) throws IOException, InterruptedException {
            return get("/v2/2finance-network/markets/" + encode(market) + "/candles" + (query == null || query.isBlank() ? "" : "?" + query));
        }

        public String products(String productType) throws IOException, InterruptedException {
            return get("/v2/2finance-network/products/" + encode(productType));
        }

        public String createProduct(String productType, String jsonBody) throws IOException, InterruptedException {
            return post("/v2/2finance-network/products/" + encode(productType), jsonBody);
        }

        public String bonds() throws IOException, InterruptedException {
            return products("bonds");
        }

        public String createBond(String jsonBody) throws IOException, InterruptedException {
            return createProduct("bonds", jsonBody);
        }

        public String loans() throws IOException, InterruptedException {
            return products("loans");
        }

        public String createLoan(String jsonBody) throws IOException, InterruptedException {
            return createProduct("loans", jsonBody);
        }

        public String swaps() throws IOException, InterruptedException {
            return products("swaps");
        }

        public String createSwap(String jsonBody) throws IOException, InterruptedException {
            return createProduct("swaps", jsonBody);
        }

        public String stakingProducts() throws IOException, InterruptedException {
            return products("staking");
        }

        public String createStakingProduct(String jsonBody) throws IOException, InterruptedException {
            return createProduct("staking", jsonBody);
        }

        public String syntheticAssets() throws IOException, InterruptedException {
            return products("synthetic-assets");
        }

        public String createSyntheticAsset(String jsonBody) throws IOException, InterruptedException {
            return createProduct("synthetic-assets", jsonBody);
        }

        public String liquidityPools() throws IOException, InterruptedException {
            return products("liquidity-pools");
        }

        public String createLiquidityPool(String jsonBody) throws IOException, InterruptedException {
            return createProduct("liquidity-pools", jsonBody);
        }
    }

    public static class TradingControlClient extends ServiceClient {
        public TradingControlClient(String baseUrl, HttpClient httpClient, TokenSource tokenSource) {
            super(baseUrl, httpClient, tokenSource);
        }

        public String robots() throws IOException, InterruptedException {
            return get("/robots");
        }

        public String createRobot(String jsonBody) throws IOException, InterruptedException {
            return post("/robots", jsonBody);
        }

        public String robot(String id) throws IOException, InterruptedException {
            return get("/robots/" + encode(id));
        }

        public String startRobot(String id) throws IOException, InterruptedException {
            return post("/robots/" + encode(id) + ":start", null);
        }

        public String pauseRobot(String id) throws IOException, InterruptedException {
            return post("/robots/" + encode(id) + ":pause", null);
        }

        public String resumeRobot(String id) throws IOException, InterruptedException {
            return post("/robots/" + encode(id) + ":resume", null);
        }

        public String stopRobot(String id) throws IOException, InterruptedException {
            return post("/robots/" + encode(id) + ":stop", null);
        }

        public String riskPolicy(String id) throws IOException, InterruptedException {
            return get("/robots/" + encode(id) + "/risk-policy");
        }

        public String setRiskPolicy(String id, String jsonBody) throws IOException, InterruptedException {
            return put("/robots/" + encode(id) + "/risk-policy", jsonBody);
        }

        public String riskView(String id) throws IOException, InterruptedException {
            return get("/risk-view/" + encode(id));
        }

        public String strategies() throws IOException, InterruptedException {
            return get("/strategies");
        }

        public String createStrategy(String jsonBody) throws IOException, InterruptedException {
            return post("/strategies", jsonBody);
        }

        public String directives() throws IOException, InterruptedException {
            return get("/directives");
        }

        public String createDirective(String jsonBody) throws IOException, InterruptedException {
            return post("/directives", jsonBody);
        }

        public String audit() throws IOException, InterruptedException {
            return get("/audit");
        }

        public String activity() throws IOException, InterruptedException {
            return get("/activity");
        }

        public String mcpTools() throws IOException, InterruptedException {
            return get("/mcp/tools");
        }
    }

    public static class KeyStoreClient extends ServiceClient {
        public KeyStoreClient(String baseUrl, HttpClient httpClient, TokenSource tokenSource) {
            super(baseUrl, httpClient, tokenSource);
        }

        public String health() throws IOException, InterruptedException {
            return get("/healthz");
        }

        public String readiness() throws IOException, InterruptedException {
            return get("/readyz");
        }

        public String startKeygen(String jsonBody) throws IOException, InterruptedException {
            return post("/keystore/keygen/start", jsonBody);
        }

        public String keygenSignature(String jsonBody) throws IOException, InterruptedException {
            return post("/keystore/keygen/signature", jsonBody);
        }

        public String startSigning(String jsonBody) throws IOException, InterruptedException {
            return post("/keystore/signing/start", jsonBody);
        }

        public String signingSignature(String jsonBody) throws IOException, InterruptedException {
            return post("/keystore/signing/signature", jsonBody);
        }

        public String startResharing(String jsonBody) throws IOException, InterruptedException {
            return post("/keystore/resharing/start", jsonBody);
        }

        public String keys(String userPublicKey) throws IOException, InterruptedException {
            return get("/keystore/keys/" + encode(userPublicKey));
        }

        public String signatures(String userPublicKey) throws IOException, InterruptedException {
            return get("/keystore/signatures/" + encode(userPublicKey));
        }

        public String metrics() throws IOException, InterruptedException {
            return get("/keystore/tss/metrics");
        }
    }

    public static class HummingbotClient extends ServiceClient {
        public HummingbotClient(String baseUrl, HttpClient httpClient, TokenSource tokenSource) {
            super(baseUrl, httpClient, tokenSource);
        }

        public String assets() throws IOException, InterruptedException {
            return get("/api/v1/assets");
        }

        public String symbols() throws IOException, InterruptedException {
            return get("/api/v1/symbols");
        }

        public String balances() throws IOException, InterruptedException {
            return get("/api/v1/balances");
        }

        public String connectorConfig(String requestJson) throws IOException, InterruptedException {
            return post("/api/v1/connectors/2finance/config", requestJson);
        }
    }

    public static class ProviderClient extends ServiceClient {
        public ProviderClient(String baseUrl, HttpClient httpClient, TokenSource tokenSource) {
            super(baseUrl, httpClient, tokenSource);
        }
    }

    public static class WiseClient extends ProviderClient {
        public WiseClient(String baseUrl, HttpClient httpClient, TokenSource tokenSource) {
            super(baseUrl, httpClient, tokenSource);
        }

        public String profiles() throws IOException, InterruptedException {
            return get("/v1/profiles");
        }

        public String profile(String profileId) throws IOException, InterruptedException {
            return get("/v1/profiles/" + encode(profileId));
        }

        public String createQuote(String profileId, String jsonBody) throws IOException, InterruptedException {
            return post("/v3/profiles/" + encode(profileId) + "/quotes", jsonBody);
        }

        public String createTransfer(String jsonBody) throws IOException, InterruptedException {
            return post("/v1/transfers", jsonBody);
        }
    }

    public static class AirwallexClient extends ProviderClient {
        public AirwallexClient(String baseUrl, HttpClient httpClient, TokenSource tokenSource) {
            super(baseUrl, httpClient, tokenSource);
        }

        public String accounts() throws IOException, InterruptedException {
            return get("/api/v1/accounts");
        }

        public String payments() throws IOException, InterruptedException {
            return get("/api/v1/payments");
        }

        public String createPayment(String jsonBody) throws IOException, InterruptedException {
            return post("/api/v1/payments", jsonBody);
        }

        public String beneficiaries() throws IOException, InterruptedException {
            return get("/api/v1/beneficiaries");
        }

        public String createBeneficiary(String jsonBody) throws IOException, InterruptedException {
            return post("/api/v1/beneficiaries", jsonBody);
        }
    }

    public static class MatchEngineClient {
        public final String webSocketUrl;

        public MatchEngineClient(String webSocketUrl) {
            this.webSocketUrl = webSocketUrl == null ? "" : webSocketUrl.trim();
        }

        public String orderCommand(String commandJson) {
            String body = commandJson == null || commandJson.isBlank() ? "{}" : commandJson.trim();
            if (body.equals("{}")) {
                return "{\"schema\":\"matchengine.order_command.v2\",\"message_type\":\"ORDER\",\"operation\":\"ADD\"}";
            }
            return "{\"schema\":\"matchengine.order_command.v2\",\"message_type\":\"ORDER\",\"operation\":\"ADD\"," + body.replaceFirst("^\\{", "");
        }

        public String marketDataSubscribe(String requestJson) {
            String body = requestJson == null || requestJson.isBlank() ? "{}" : requestJson.trim();
            if (body.equals("{}")) {
                return "{\"schema\":\"matchengine.market_data_subscribe.v2\"}";
            }
            return "{\"schema\":\"matchengine.market_data_subscribe.v2\"," + body.replaceFirst("^\\{", "");
        }

        public String sendOrder(Function<String, String> sender, String commandJson) {
            return sender.apply(orderCommand(commandJson));
        }

        public String subscribeMarketData(Function<String, String> sender, String requestJson) {
            return sender.apply(marketDataSubscribe(requestJson));
        }
    }

    public static class PlannerClient {
        private final MCPClient mcp;
        private final OrchestratorClient orchestrator;
        private final AnalyticsClient analytics;
        private final TradingControlClient tradingControl;

        public PlannerClient(
                MCPClient mcp,
                OrchestratorClient orchestrator,
                AnalyticsClient analytics,
                TradingControlClient tradingControl
        ) {
            this.mcp = mcp;
            this.orchestrator = orchestrator;
            this.analytics = analytics;
            this.tradingControl = tradingControl;
        }

        public String conversationPlan(String argumentsJson) throws IOException, InterruptedException {
            return mcp.conversationPlan(argumentsJson);
        }

        public String orchestratedPlan(String jsonBody) throws IOException, InterruptedException {
            return orchestrator.sendMessage(jsonBody);
        }

        public String operationalPlan(String jsonBody) throws IOException, InterruptedException {
            return orchestratedPlan(jsonBody);
        }

        public String tradingPlan(String requestJson, boolean useAnalytics, boolean useTrading)
                throws IOException, InterruptedException {
            StringBuilder context = new StringBuilder("{");
            boolean[] hasContext = {false};
            if (useTrading && tradingControl != null) {
                try {
                    appendContextValue(context, hasContext, "trading_robots", tradingControl.robots());
                } catch (IOException | InterruptedException error) {
                    if (error instanceof InterruptedException) {
                        Thread.currentThread().interrupt();
                    }
                }
            }
            if (useAnalytics && analytics != null) {
                try {
                    appendContextValue(context, hasContext, "analytics_indicators", analytics.indicators());
                } catch (IOException | InterruptedException error) {
                    if (error instanceof InterruptedException) {
                        Thread.currentThread().interrupt();
                    }
                }
            }
            context.append("}");
            return conversationPlan("{\"request\":" + jsonValueOrString(requestJson) + ",\"context\":" + context + "}");
        }

        private static void appendContextValue(
                StringBuilder context,
                boolean[] hasContext,
                String key,
                String value
        ) {
            if (hasContext[0]) {
                context.append(",");
            }
            context.append("\"").append(escape(key)).append("\":").append(jsonValueOrString(value));
            hasContext[0] = true;
        }

        private static String jsonValueOrString(String value) {
            String trimmed = value == null ? "" : value.trim();
            if (trimmed.isEmpty()) {
                return "null";
            }
            char first = trimmed.charAt(0);
            if (first == '{'
                    || first == '['
                    || first == '"'
                    || first == '-'
                    || Character.isDigit(first)
                    || trimmed.equals("true")
                    || trimmed.equals("false")
                    || trimmed.equals("null")) {
                return trimmed;
            }
            return "\"" + escape(trimmed) + "\"";
        }
    }

    private static String encode(String value) {
        return URLEncoder.encode(value == null ? "" : value, StandardCharsets.UTF_8).replace("+", "%20");
    }

    private static String escape(String value) {
        return (value == null ? "" : value).replace("\\", "\\\\").replace("\"", "\\\"");
    }
}
